doc: Add ISE snapshot doc for H20 (#28963)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54h/ug_nrf54h20_ironside_snapshot.rst

Overview
********

Snapshot services capture a configured set of MRAM regions, encrypt the captured data, and store the encrypted data in external memory.
When corruption is detected in a captured region, |ISE| restores the affected MRAM content from the encrypted snapshot.

.. note::
   Snapshot services depend on device support, |ISE| support, and production-time configuration.
   For information about checking support on a device, see :ref:`ug_nrf54h20_ironside_se_snapshot_identify_support`.

.. _ug_nrf54h20_ironside_se_snapshot_workflow:

Snapshot workflow
*****************

Snapshot setup is a production workflow.
Configure snapshot regions before locking the UICR, and capture the first snapshot after the UICR is locked.

The following workflow shows the typical order of operations:

1. Identify the critical MRAM regions that must be recoverable.
#. Configure snapshot regions in UICR during production.
#. Lock the UICR configuration.
#. Invoke the |ISE| snapshot capture service from local-domain firmware.
#. Check the :ref:`boot report <ug_nrf54h20_ironside_se_boot_report>` after the device resets.

> #. During operation, handle either automatic snapshot recovery or an |ISE| event report.

Reviewer: Is this telling the user to handle automatic snapshot recovery or an ISE event report?
The next section talks about runtime as well, maybe it'd be easier to mark the first list as setup configuration and the next one as runtime behavior?

During runtime, snapshot handling follows this flow:

1. The device boots normally after a successful capture.
#. If MRAMC detects corruption during an MRAM read, |ISE| handles the MRAMC interrupt.
#. If the corrupted address is inside a captured snapshot region, |ISE| requests automatic snapshot recovery from SDROM and resets the device.
#. If the corrupted address is outside captured snapshot regions, |ISE| reports the address through the event report.

> #. Local-domain firmware also runs integrity checks before reading MRAM content and requests snapshot recovery when those checks fail.

Reviewer: ```suggestion
#. Local-domain firmware can run integrity checks before reading MRAM content and requests snapshot recovery when those checks fail.
```

> #. Local-domain firmware reads the boot report or event report and handles any application-specific recovery.

Reviewer: ```suggestion
#. Local-domain firmware can read the boot report or event report and handles any application-specific recovery.
```

The MRAMC interrupt is connected to and handled by |ISE|.
When |ISE| handles the interrupt, it checks whether the corrupted address belongs to a captured snapshot region.

Snapshot region handling
========================

|ISE| handles detected corruption differently depending on whether the corrupted address is covered by snapshot:

* If the address is in a captured snapshot region, |ISE| requests snapshot recovery from SDROM and resets the device without local-domain involvement.
* If the address is outside the captured snapshot regions, |ISE| reports the corrupted address to local domains through the |ISE| event report.

The ability to recover the snapshot content is limited to the regions that were included in the snapshot capture.
Local domains remain responsible for recovering corrupted content outside snapshot regions.

Application integrity checks
============================

MRAMC ECC cannot detect every corruption scenario.
Any firmware that reads MRAM content must run integrity checks on MRAM before reading it and request snapshot recovery if the integrity check fails.

> SysCtrl, SDROM, and |ISE| run integrity checks for the MRAM content that they read.

Reviewer 2: SysCtrl ROM?

Together, these checks verify the integrity of the following areas:

* FICR
* RICR
* SICR
* BICR
* UICR

> * UROT

Reviewer 2: Maybe this should just state the IronSide SE firmware?

* ``UICR.PROTECTEDMEM``

> * ``UICR.SECURESTORAGE``

Reviewer 2: Should this state something like "the area defined by UICR.x" to make it clear that it is not the UICR field itself?

Also note that it checks either UICR.PROTECTEDMEM or UICR.SECONDARY.PROTECTEDMEM depending on whether secondary is booted

* |ISE| internal storage

.. _ug_nrf54h20_ironside_se_snapshot_regions:

Snapshot region configuration
*****************************

Snapshot can capture up to 16 MRAM regions.
Nine regions are predefined by |ISE|, and up to seven regions are configurable during production.

Region limits
=============

All snapshot regions must meet the following requirements:

* The region start address must be aligned to 1 KiB.

> * The region size must be aligned to 1 KiB.

Reviewer 2: ```suggestion
* The region start address must be aligned to 2 KiB.
* The region size must be aligned to 2 KiB.
```

* Configurable regions must not overlap each other.
* Configurable regions must not overlap the predefined regions.
* A region must not cross the MRAM10 and MRAM11 boundary at ``0x0E10_0000``.
* Configurable regions can be configured only once during production.

Predefined regions
==================

Eight predefined regions cover the four 2 KiB NVR pages, also known as information pages, for each MRAM instance:

* MRAM10
* MRAM11

The ninth predefined region covers SICR and |ISE| firmware.
This region starts at the beginning of MRAM10, ``0x0E00_0000``, and continues until, but does not include, the immutable bootloader for the application core at ``0x0E03_0000``.

Configurable regions
====================

Configure the seven application-defined snapshot regions in UICR during production.

> For general information about UICR configuration, see :ref:`ug_nrf54h20_ironside_se_uicr`.

Reviewer: The uicr description doesn't have the fields particular to snapshot

At minimum, include the memory protected by :ref:`UICR.PROTECTEDMEM <ug_nrf54h20_ironside_se_uicr_protectedmem>` in a configurable snapshot region.
This memory starts at ``0x0E03_0000`` and typically contains the immutable bootloader for the application core.

> Place ``UICR.PERIPHCONF`` data in the ``UICR.PROTECTEDMEM`` area.

Reviewer 2: UICR.PERIPHCONF and UICR.MPCCONF data?

If secure storage is enabled, also include the memory configured by :ref:`UICR.SECURESTORAGE <ug_nrf54h20_ironside_uicr_securestorage>` in a configurable snapshot region.

> Secure storage does not need a snapshot region only when secure storage is disabled.

Reviewer 2: The phrasing here is a bit confusing, maybe the sentence could just be removed? If I understood it correctly it states more or less the same as the one above.

Recommended region strategies
=============================

Choose snapshot regions based on the content that must be recoverable after MRAM corruption.
Common strategies include the following:

* Capture only required immutable and secure regions, such as ``UICR.PROTECTEDMEM`` and ``UICR.SECURESTORAGE``.
* Cover all MRAM with the combined predefined and configurable regions to maximize recoverability.
* Cover all MRAM except areas that contain file systems.

When a region is not covered by snapshot, you must provide your own recovery strategy.
For example, corruption in a file system might require reformatting, while corruption in firmware might require device firmware update (DFU).

.. _ug_nrf54h20_ironside_se_snapshot_external_memory:

External memory requirements
****************************

> Snapshot stores encrypted copies of captured regions in external memory.

Reviewer: This doesn't say anything on how to configure the external flash module, so what pins it is connected, what type of module, frequency and what not. I suppose this is done in a different document and that you get the device ready for this, but this should still be mentioned somewhere

The external memory chip used by snapshot must have a page size of 4 KiB.

Use the following formula to calculate the required external memory capacity:

.. code-block:: text

   2 * (4 * 1024 + sum(round_up(region.size, 4 * 1024) for region in snapshot.regions))

In this formula:

* ``region.size`` is the size of one configured snapshot region in bytes.
* ``round_up(region.size, 4 * 1024)`` rounds each region up to a 4 KiB boundary.

> * ``snapshot.regions`` is the complete set of predefined and configurable snapshot regions.

Reviewer 2: If the idea is to apply this formula by hand then I think we should simplify it a bit and combine the constant parts into one so that the user doesn't have to do that calculation.
Looks like the predefined regions take up `229376` bytes here (8 regions of size 0x800 and one with size 0x30000).
Combining that with the other constant parts I get a constant part of 466944 = 2 * (4096 + 229376), so we could make the formula something like:

```
2 * sum(round_up(region.size, 4096) for region in snapshot.regions) + 466944
```
and state that:

> ``snapshot.regions`` are the configurable snapshot regions.

.. _ug_nrf54h20_ironside_se_snapshot_configure_regions:

Configure snapshot regions
**************************

Configure snapshot regions when you prepare the device for production.
The configured regions determine which MRAM content |ISE| can restore during snapshot recovery.

Before configuring snapshot regions, ensure the following:

* You have planned the ``UICR.PROTECTEDMEM`` and ``UICR.SECURESTORAGE`` locations.
* You have confirmed that the selected regions satisfy the snapshot alignment and overlap requirements.

.. caution::
   Configure snapshot regions carefully before locking UICR.
   Configurable snapshot regions can be configured only once during production.

To configure snapshot regions, complete the following steps:

1. Add the memory protected by ``UICR.PROTECTEDMEM`` to a configurable snapshot region.
#. If secure storage is enabled, add the ``UICR.SECURESTORAGE`` memory to a configurable snapshot region.
#. Add any other application-specific MRAM regions that must be recoverable.
#. Verify that no configurable region overlaps another configurable region or a predefined region.

> #. Verify that no configurable region crosses the MRAM10 and MRAM11 boundary at ``0x0E10_0000``.

Reviewer 2: We describe the requirements for the configurable regions pretty extensively in a section above, so maybe this part should just reference that instead of listing them out again? We also say above here that the user should ensure correctness before doing this, so maybe these steps could just be dropped here instead.

* The snapshot regions are configured.
* The UICR is locked.
* The device has sufficient power for the complete capture operation.
* The device is in an environment where MRAM corruption is not expected during capture.

.. caution::
   Do not capture a snapshot when power loss or MRAM corruption can occur.
   The capture operation is not resistant to power loss or magnetic corruption.

To capture a snapshot, complete the following steps:

1. Boot the device.
   Capture shortly after boot because the boot process detects most corruption scenarios.

#. From local-domain firmware, invoke the |ISE| snapshot capture service.
   The device resets during the capture operation.

#. After reset, read the :ref:`boot report <ug_nrf54h20_ironside_se_boot_report>` to check the capture result.

The boot report indicates whether the capture completed successfully.

.. _ug_nrf54h20_ironside_se_snapshot_recovery:

Handle snapshot recovery
************************

Snapshot recovery is automatic when |ISE| detects recoverable corruption in a captured snapshot region.
Local domains do not need to take part in the restore operation.

> During snapshot recovery, |ISE| performs the following operations:

Reviewer 2: Since we distinguish between SDROM and ISE above here:

```suggestion
During snapshot recovery, SDROM performs the following operations:
```

> 1. Performs a hardware ERASEALL operation for the MRAM10 bank.

Reviewer: I'd highlight in a note that the device gets erased and only snapshot regions are restored, so everything outside of snapshot regions will be lost.

The snapshot mechanism does not recover MRAM outside the captured regions.

Typical recovery actions include the following:

* Reformat a corrupted file system.
* Trigger DFU when firmware corruption is detected.

.. _ug_nrf54h20_ironside_se_snapshot_identify_support:

Identify snapshot support
*************************

You can identify snapshot support in two ways:

* In lifecycle state ``EMPTY``, read FICR directly.
* In lifecycle state ``RoT`` or ``DEPLOYED``, check the |ISE| boot report in RAM.

Identify support in lifecycle state EMPTY
=========================================

To identify whether an nRF54H20 device in lifecycle state ``EMPTY`` supports the |ISE| snapshot service, read the FICR directly:

.. code-block:: console

   nrfutil device read --traits jlink --core secure --direct --address 0x0FFFE054

If the value is ``0x0005420B``, the device supports the |ISE| snapshot service.

Identify support in lifecycle state RoT or DEPLOYED
===================================================

To identify whether a device in lifecycle state ``RoT`` or ``DEPLOYED`` supports snapshot, check the |ISE| boot report in RAM.

> For more information about the boot report, see :ref:`ug_nrf54h20_ironside_se_boot_report`.

Reviewer: The ug_nrf54h20_ironside_se_boot_report reference doesn't say how to find out if the snapshot feature is supported or not. Would it make sense to reference the header file instead? It has fairly good documentation

Author: @karstenkoenig specifically, which header file available in the sdk-nrf repo do you have in mind?

Reviewer: https://github.com/zephyrproject-rtos/hal_nordic/blob/master/ironside/se/include/ironside/se/boot_report.h#L219-L245
